lib/Doctrine/Common/Collections/Collection.php

```diff
@@ -212,6 +212,19 @@ public function next();
      */
     public function exists(Closure $p) : bool;
 
+    /**
+     * Returns the first element of this collection that satisfies the predicate p.
+     *
+     * @param Closure $p The predicate.
+     *
+     * @return mixed The first element respecting the predicate,
+     *               null if no element respects the predicate.
+     *
+     * @psalm-param Closure(TKey=, T=):bool $p
+     * @psalm-return T|null
+     */
+    public function findOne(Closure $p);
```

Comment:
I'm wondering how was new `findOne` method chosen. Name doesn't indicate it will return first matching result. For all I know, it can return last matching result. In fact, it's doing same thing as `first()`, except it skips non-matching results. I'm especially weirded out by this choice, because even laravel/collections uses `first(callable)` for this.

Reply:
I previously named this `find`. Then @greg0ire asked for `findOne` https://github.com/doctrine/collections/pull/252#discussion_r475087688.

I don't think as weird the term `find`. It seems familiar to me.
This is similar to JS: https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Array/find
Or to doctrine/ORM `findOne` method.

IMHO, the fact that this is the *first* matching result should be something relevant.
I introduced this method because
- You may need to know that the element exists AND you need to use it.
- Using both `filter` and `first` are under-performant for this case.

Reply:
JS doesn't already have `first` function though. ORM does have `find` function, but it has completely different signature than this one. In fact, without knowing what these functions do beforehand, I wouldn't be able to tell the difference between `Collection::findOne` and `ObjectRepository::find` based on name, or I would think `ObjectRepository::find` returns more results than `Collection::findOne`, then be confused why one expects ID and another one closure.

@greg0ire ok to switch back to `first`?

Reply:
> @greg0ire ok to switch back to `first`?

`first` already exist, it
```
Sets the internal iterator to the first element in the collection and returns this element.
```

The method doesn't modify the internal iterator, and should not.
I don't think it's a good idea to modify the signature from `first()` to `first(?callable)`.

A specific function seems better.

> ORM does have find function, but it has completely different signature than this one. In fact, without knowing what these functions do beforehand, I wouldn't be able to tell the difference between Collection::findOne and ObjectRepository::find based on name, or I would think ObjectRepository::find returns more results than Collection::findOne, then be confused why one expects ID and another one closure.

I don't know why you compare to `find`, you should compare to `findOne` instead.
Both `Collection::findOne` and `ObjectRepository::findOne` will return 1 value.

Reply:
> I don't know why you compare to find, you should compare to findOne instead.

`ObjectRepository::findOne` doesn't exist either. Are you thinking of `findOneBy`?

> The method doesn't modify the internal iterator, and should not.

This is good argument, but TBH I don't see the utility of functions like current(), next() and key() or other ways to play with internal pointer. I never saw a use for those, it was always foreach. We might need to think of removing them. Then nobody would have issues related to internal pointer, as it wouldn't be exposed.

> IMHO, the fact that this is the first matching result should be something relevant.

WDYM? This is why I suggested to use `first`. I don't complain about `One` part, I'm on board with that. Otherwise `findOne` name doesn't tell it returns first matching result. Yes, it's consistent with current implicit Persistence API, doesn't mean we can't create better, more explicit API going forward. Perhaps `findFirst()`?

Reply:
> > I don't know why you compare to find, you should compare to findOne instead.
> 
> `ObjectRepository::findOne` doesn't exist either. Are you thinking of `findOneBy`?

Yes indeed ! I was thinking of `findOneBy`, sorry for my mistake.

> > The method doesn't modify the internal iterator, and should not.
> 
> This is good argument, but TBH I don't see the utility of functions like current(), next() and key() or other ways to play with internal pointer. I never saw a use for those, it was always foreach. We might need to think of removing them. Then nobody would have issues related to internal pointer, as it wouldn't be exposed.

I never played with iterator either.

> > IMHO, the fact that this is the first matching result should be something relevant.
> 
> WDYM?

I was kinda asking myself if the fact that it was the first matching result was something important.

For instance if I had something as 
```
new ArrayCollection(
    ['id' => 1, 'ready' => true],
    ['id' => 2, 'ready' => false],
    ['id' => 3, 'ready' => false],
    ['id' => 4, 'ready' => true],
    ['id' => 5, 'ready' => true],
)
```
And I look for a slot which is ready, I won't care if it's the id `1`, `4` or `5`.
I also don't care if sometimes this configuration gives me `1` and sometimes it gives me `5`. I just want one.

If the fact that it's the `first` result is not relevant, `findOne` seems ok.
And then maybe the doc should be updated to
```
Returns one element of this collection that satisfies the given predicate.
```

If the fact the it's the `first` result is important, `findFirst` could be indeed better.

Reply:
Php does have a `array_search` method ; maybe `searchValue` is better ?